server/src/com/thoughtworks/go/config/update/UpdatePackageConfigCommand.java

```diff
@@ -0,0 +1,120 @@
+    private final HttpLocalizedOperationResult result;
+
+    public UpdatePackageConfigCommand(GoConfigService goConfigService, String oldPackageId, PackageDefinition newPackage, Username username, String md5, EntityHashingService entityHashingService, HttpLocalizedOperationResult result, PackageDefinitionService packageDefinitionService) {
+        super(newPackage, result, packageDefinitionService);
+        this.goConfigService = goConfigService;
+        this.oldPackageId = oldPackageId;
+        this.newPackage = newPackage;
+        this.username = username;
+        this.md5 = md5;
+        this.entityHashingService = entityHashingService;
+        this.result = result;
+    }
+
+    @Override
+    public void update(CruiseConfig preprocessedConfig) throws Exception {
+        PackageRepositories packageRepositories = preprocessedConfig.getPackageRepositories();
+        PackageRepository repository = packageRepositories.findPackageRepositoryHaving(oldPackageId);
+        int index = packageRepositories.indexOf(repository);
+
+        newPackage.setRepository(repository);
+        repository.removePackage(oldPackageId);
+        repository.addPackage(this.newPackage);
+
+        packageRepositories.remove(index);
+        packageRepositories.add(repository);
+    }
+
+    @Override
+    public void clearErrors() {
+        BasicCruiseConfig.clearErrors(preprocessedPackageDefinition);
+    }
+
+    @Override
+    public boolean isValid(CruiseConfig preprocessedConfig) {
+        return super.isValid(preprocessedConfig, newPackage.getRepository().getId());
+    }
+
+    @Override
+    public PackageDefinition getPreprocessedEntityConfig() {
+        return preprocessedPackageDefinition;
+    }
+
+    @Override
+    public boolean canContinue(CruiseConfig cruiseConfig) {
+        return isAuthorized() && isRepositoryPresent(cruiseConfig) && isIdSame() && isRequestFresh();
+    }
+
+    private boolean isRepositoryPresent(CruiseConfig cruiseConfig) {
+        String repoId = newPackage.getRepository().getRepoId();
+        if (cruiseConfig.getPackageRepositories().find(repoId) == null) {
+            result.unprocessableEntity(LocalizedMessage.string("PACKAGE_REPOSITORY_NOT_FOUND", repoId));
+            return false;
+        }
+        return true;
+    }
+
+    private boolean isAuthorized() {
+        if (!(goConfigService.isUserAdmin(username) || goConfigService.isGroupAdministrator(username.getUsername()))) {
+            result.unauthorized(LocalizedMessage.string("UNAUTHORIZED_TO_EDIT"), HealthStateType.unauthorised());
+            return false;
+        }
+        return true;
+    }
+
+    private boolean isIdSame() {
+        boolean isPackageIdSame = newPackage.getId().equals(oldPackageId);
+        if (!isPackageIdSame) {
+            result.unprocessableEntity(LocalizedMessage.string("PACKAGE_REPOSITORY_ID_CHANGED", "package"));
+        }
+        return isPackageIdSame;
+    }
+
+    private boolean isRequestFresh() {
```

Comment:
Sorry — I must have marked the wrong method. That's supposed to be on the isAuthorized method.